packages/server-core/src/media/static-resource/static-resource.hooks.ts

```diff
@@ -69,6 +76,45 @@ const ensureResource = async (context: HookContext<StaticResourceService>) => {
   }
 }
 
+const createHashIfNeeded = async (context: HookContext<StaticResourceService>) => {
+  if (!context.data || !(context.method === 'create' || context.method === 'update')) {
+    throw new BadRequest(`${context.path} service only works for data in ${context.method}`)
+  }
+
+  if (Array.isArray(context.data)) throw new BadRequest('Batch create is not supported')
+
+  const data = context.data
+
+  if (!data.key) throw new BadRequest('key is required')
+
+  if (!data.hash) {
+    const storageProvider = getStorageProvider()
+    const [_, directory, file] = /(.*)\/([^\\\/]+$)/.exec(data.key)!
+    if (!(await storageProvider.doesExist(file, directory))) throw new BadRequest('File could not be found')
+    const result = await storageProvider.getObject(data.key)
+    const hash = createStaticResourceHash(result.Body)
+    context.data.hash = hash
+  }
+}
+
+const updateResourcesJson = async (context: HookContext<StaticResourceService>) => {
+  if (!context.result) throw new BadRequest('Batch update is not supported')
+  const ignoreResourcesJson = context.params?.ignoreResourcesJson
+  if (ignoreResourcesJson) return
+
+  const results =
+    'data' in context.result ? context.result.data : Array.isArray(context.result) ? context.result : [context.result]
+
+  const projectNames = results.reduce((acc: string[], result) => {
+    if (result.project && result.key.startsWith('projects/') && !acc.includes(result.project)) acc.push(result.project)
+    return acc
+  }, [])
+
+  for (const projectName of projectNames) {
+    await updateProjectResourcesJson(context.app, projectName)
```

Comment:
We shouldn't fully regenerate the resources.json with every CRUD operation. We need specific patch and delete functions which perform singular operations on the resource.json

Reply:
^ seems like a reasonable optimization

Reply:
I actually disagree here. Batch uploads can/should be supported at the service level when needed, and that kind of optimisation I don't feel is appropriate at this stage. The difference between a partial write and a complete write is negligible, as the query/read/write requests themselves are the heavy part, not the size of the query.